llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -716,6 +716,18 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
       return replaceInstUsesWith(I, Pow);
     }
 
+    // powi(X, Y) * X --> powi(X, Y+1)
+    // X * powi(X, Y) --> powi(X, Y+1)
+    if (match(&I, m_c_FMul(m_OneUse(m_Intrinsic<Intrinsic::powi>(m_Value(X),
```

Comment:
The reassoc check is really hard to find, this whole if (reassoc) should probably be split to a helper function.

Also, should check the reason on the inner operation too?

Reply:
hi @arsenm , do you mean both the operation `fmul `and `powi `should have the `reassoc ` ?
  I find all the past combines only check the instruction `I` itself only.

Reply:
I consider that to be a bug. I think we should have strict flag requirements for both participants

Reply:
Fixed with new commit 4423e80cb, thanks

Reply:
This commit isn't upstream, or in a PR? Did you mean to link a PR with it?

Reply:
oh, sorry, please ignore that. I have rebase it on this PR (commit 86b9575) because C/C++ code formatter, and forget to update the reply.